File: src/main/java/com/baulsupp/oksocial/services/facebook/FacebookApiDocPresenter.java
package com.baulsupp.oksocial.services.facebook;

import com.baulsupp.oksocial.apidocs.ApiDocPresenter;
import com.baulsupp.oksocial.authenticator.oauth2.Oauth2Token;
import com.baulsupp.oksocial.credentials.ServiceDefinition;
import com.google.common.base.Throwables;
import com.baulsupp.oksocial.output.OutputHandler;
import java.io.IOException;
import java.io.InterruptedIOException;
import java.util.concurrent.ExecutionException;
import okhttp3.HttpUrl;
import okhttp3.OkHttpClient;

import static java.util.stream.Collectors.joining;

public class FacebookApiDocPresenter implements ApiDocPresenter {
  private ServiceDefinition<Oauth2Token> sd;

  public FacebookApiDocPresenter(ServiceDefinition<Oauth2Token> sd) {
    this.sd = sd;
  }

  @Override public void explainApi(String url, OutputHandler outputHandler, OkHttpClient client)
      throws IOException {
    outputHandler.info("service: " + sd.shortName());
    outputHandler.info("name: " + sd.serviceName());
    sd.apiDocs().ifPresent(d -> outputHandler.info("docs: " + d));
    sd.accountsLink().ifPresent(d -> outputHandler.info("apps: " + d));

    try {
      FacebookMetadata md = FacebookUtil.getMetadata(client, HttpUrl.parse(url)).get();
      outputHandler.info("");
      outputHandler.info("fields: " + md.fieldNames().stream().collect(joining(",")));
      outputHandler.info("");
      outputHandler.info("connections: " + md.connections().stream().collect(joining(",")));
    } catch (InterruptedException e) {
      throw (InterruptedIOException) new InterruptedIOException().initCause(e);
    } catch (ExecutionException e) {
      Throwables.propagateIfPossible(e.getCause(), IOException.class);
      throw Throwables.propagate(e.getCause());
    }
  }
}

File: src/main/kotlin/com/baulsupp/oksocial/services/facebook/FacebookApiDocPresenter.kt
package com.baulsupp.oksocial.services.facebook

import com.baulsupp.oksocial.apidocs.ApiDocPresenter
import com.baulsupp.oksocial.authenticator.oauth2.Oauth2Token
import com.baulsupp.oksocial.credentials.ServiceDefinition
import com.baulsupp.oksocial.output.OutputHandler
import okhttp3.HttpUrl
import okhttp3.OkHttpClient
import okhttp3.Response

class FacebookApiDocPresenter(private val sd: ServiceDefinition<Oauth2Token>) : ApiDocPresenter {

  override suspend fun explainApi(url: String, outputHandler: OutputHandler<Response>,
                                  client: OkHttpClient) {
    outputHandler.info("service: " + sd.shortName())
    outputHandler.info("name: " + sd.serviceName())
    sd.apiDocs()?.let { outputHandler.info("docs: " + it) }
    sd.accountsLink()?.let { outputHandler.info("apps: " + it) }

    val parsedUrl = HttpUrl.parse(url)
    // TODO handle null
    val md = FacebookUtil.getMetadata(client, parsedUrl!!)
    if (md == null) {
      outputHandler.info("")
      outputHandler.info("No metadata available")
    } else {
      outputHandler.info("")
      outputHandler.info("fields: " + md.fields.joinToString(",") { it.name })
      outputHandler.info("")
      outputHandler.info("connections: " + md.connections.keys.joinToString(","))
    }
  }
}
